File: Euler2.js
var fibgen = {
    a: 1,
    b: 1,
    next: function() {
        var temp;
        temp = this.a;
        this.a = this.b;
        this.b = this.a + temp;

        return this.a;
    }
}

function solve(upto) {
    var fibs = [];
    var tempfib = fibgen.next();
    while(tempfib < upto) {
        fibs.push(tempfib);
        tempfib = fibgen.next();
    }

    var evenfibs = fibs.filter(function(x){return x % 2 === 0});

    return evenfibs.reduce(function(x, y){return x + y});
}

console.log(solve(4000000));

File: euler2.py
from itertools import takewhile
from functools import reduce

# Solving project euler part two
#
#
def fibonacci():
    """ A generator that generates fibonacci numbers """
    a, b = 1, 2
    while True:
        yield a
        a, b = b, a + b

def solve(upto):
    """ Generates fibonacci numbers up to the limit, and then takes the sum of the even ones """
    fibgen = fibonacci()
    fibsuptogen = takewhile(lambda x: x < upto, fibgen)
    even_fibs = [ fib for fib in fibsuptogen if fib % 2 == 0]
    return reduce(lambda x, y: x + y, even_fibs)

print(solve(4000000))
